File: scripts/projections/detail.py
"""data.detail/<padded>.json projection — per DATA_ARCHITECTURE.md §6.2 (revised v1.2.0).

Status: Planned
Consumer: build_occupations.py, api/og.tsx, mobile ④/⑤ 詳細 drill-down fetch
Shape: nested object — main occupation + stats + latest score + top-N skills
Size target: < 5 KB gzipped per file

`*_top_N` rule per §6.2:
    Sort by the occupation's score descending; take first N.
    When the parent block is None (no numeric profile), the top_N field is also None.
    N: skills=10, knowledge=5, abilities=5.

v1.2.0 — EN UI removal:
    All `*_en` / `.en` fields dropped from the payload. Source occupation files
    and translation backups still carry the EN data; this projection just stops
    emitting them.

v1.1.0 additive — `sector` block:
    {
      id           : sector slug (or "_uncategorized")
      ja           : display label
      hue          : 'safe'|'mid'|'warm'|'risk' fallback tint
      provenance   : 'override'|'auto'|'auto-ambiguous'|'unmatched'|'no-mhlw'
    }
    Plus `risk_band` / `workforce_band` / `demand_band` at the top level for
    consistency with treemap + search.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING

from lib.bands import demand_band, risk_band, workforce_band  # type: ignore[import-not-found]

if TYPE_CHECKING:
    from scripts.lib.indexes import Indexes


def _top_n(block: dict[str, float] | None, labels_dim: dict, n: int) -> list[dict] | None:
    """Sort block by score desc, take top N. Return None if block is None.

    Each item: {"key", "label_ja", "score"}
    """
    if block is None:
        return None
    sorted_items = sorted(block.items(), key=lambda kv: kv[1], reverse=True)[:n]
    out = []
    for key, score in sorted_items:
        label = labels_dim.get(key)
        out.append({
            "key": key,
            "label_ja": label.ja if label else key,
            "score": score,
        })
    return out


def build(indexes: "Indexes", dist_root: Path) -> dict:
    out_dir = dist_root / "data.detail"
    out_dir.mkdir(parents=True, exist_ok=True)

    skills_labels = indexes.labels_by_dim.get("skills", {})
    knowledge_labels = indexes.labels_by_dim.get("knowledge", {})
    abilities_labels = indexes.labels_by_dim.get("abilities", {})
    sector_by_id = {s.id: s for s in indexes.sectors}

    written = 0
    for occ_id in sorted(indexes.occ_by_id):
        occ = indexes.occ_by_id[occ_id]
        trans = indexes.trans_by_id.get(occ_id)
        stats = indexes.stats_by_id.get(occ_id)
        score = indexes.latest_score_by_occ.get(occ_id)
        assignment = indexes.sector_by_occ.get(occ_id)
        sector_def = sector_by_id.get(assignment.sector_id) if assignment else None

        sector_block: dict | None = None
        if assignment is not None:
            sector_block = {
                "id": assignment.sector_id,
                "ja": sector_def.ja if sector_def else None,
                "hue": sector_def.hue if sector_def else None,
                "provenance": assignment.provenance,
            }

        payload = {
            "id": occ_id,
            "schema_version": "1.2",
            "title": {
                "ja": occ.title_ja,
                "aliases_ja": occ.aliases_ja,
            },
            "classifications": occ.classifications.model_dump(),
            # v1.1.0 additive — sector + bands for the mobile detail page
            "sector": sector_block,
            "risk_band": risk_band(score.ai_risk if score else None),
            "workforce_band": workforce_band(stats.workers if stats else None),
            "demand_band": demand_band(stats.recruit_ratio if stats else None),
            "description": {
                "summary_ja": occ.description.summary_ja,
                "what_it_is_ja": occ.description.what_it_is_ja,
                "how_to_become_ja": occ.description.how_to_become_ja,
                "working_conditions_ja": occ.description.working_conditions_ja,
            },
            "ai_risk": {
                "score": score.ai_risk,
                "model": score.model,
                "scored_at": score.date,
                "rationale_ja": score.rationale_ja,
            } if score else None,
            "stats": stats.model_dump(exclude={"id", "schema_version"}) if stats else None,
            "skills_top10": _top_n(occ.skills, skills_labels, 10),
            "knowledge_top5": _top_n(occ.knowledge, knowledge_labels, 5),
            "abilities_top5": _top_n(occ.abilities, abilities_labels, 5),
            "tasks_count": len(occ.tasks),
            "tasks_lead_ja": occ.tasks_lead_ja,
            "related_orgs": [o.model_dump() for o in occ.related_orgs],
            "related_certs_ja": occ.related_certs_ja,
            "url": occ.url,
            "data_source_versions": occ.data_source_versions.model_dump(),
        }

        out = out_dir / f"{occ_id:04d}.json"
        out.write_text(json.dumps(payload, ensure_ascii=False, separators=(",", ":")) + "\n", encoding="utf-8")
        written += 1

    return {"dir": out_dir, "files": written}

File: src/data/projections/detail.ts
/**
 * data.detail/<padded>.json projection — per docs/DATA_ARCHITECTURE.md §6.2 (v1.2.0).
 *
 * Status: Planned
 * Consumer: build_occupations.py (legacy), api/og.tsx, mobile ④/⑤ 詳細
 * Shape: nested object — main occupation + stats + latest score + top-N skills
 *
 * `*_top_N` rule per §6.2:
 *   Sort by the occupation's score descending; take first N.
 *   When the parent block is null (no numeric profile), the top_N field is also null.
 *   N: skills=10, knowledge=5, abilities=5.
 *
 * Migrated from scripts/projections/detail.py.
 */
import { mkdir, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import type { Indexes } from '../lib/indexes.js';
import type { LabelEntry } from '../schema/labels.js';
import { demandBand, riskBand, workforceBand } from '../lib/bands.js';

interface TopNEntry {
  key: string;
  label_ja: string;
  score: number;
}

export interface DetailBuildResult {
  files: string[];
  dir: string;
  fileCount: number;
}

/** Sort `block` by score descending, take top `n`. Return null when block is null. */
function topN(
  block: Record<string, number> | null | undefined,
  labelsDim: Map<string, LabelEntry>,
  n: number,
): TopNEntry[] | null {
  if (block == null) return null;
  // Python's `sorted(items, key=..., reverse=True)` is stable.
  // V8's Array.prototype.sort is stable since Node 12.
  const entries = Object.entries(block).sort((a, b) => b[1] - a[1]);
  const out: TopNEntry[] = [];
  for (const [key, score] of entries.slice(0, n)) {
    const label = labelsDim.get(key);
    out.push({
      key,
      label_ja: label ? label.ja : key,
      score,
    });
  }
  return out;
}

export async function buildDetail(
  indexes: Indexes,
  distRoot: string,
): Promise<DetailBuildResult> {
  const outDir = join(distRoot, 'data.detail');
  await mkdir(outDir, { recursive: true });

  const skillsLabels = indexes.labelsByDim.get('skills') ?? new Map<string, LabelEntry>();
  const knowledgeLabels = indexes.labelsByDim.get('knowledge') ?? new Map<string, LabelEntry>();
  const abilitiesLabels = indexes.labelsByDim.get('abilities') ?? new Map<string, LabelEntry>();
  const sectorById = new Map(indexes.sectors.map((s) => [s.id, s] as const));

  const writtenFiles: string[] = [];
  const sortedIds = [...indexes.occById.keys()].sort((a, b) => a - b);

  for (const occId of sortedIds) {
    const occ = indexes.occById.get(occId)!;
    const stats = indexes.statsById.get(occId);
    const score = indexes.latestScoreByOcc.get(occId);
    const assignment = indexes.sectorByOcc.get(occId);
    const sectorDef = assignment ? sectorById.get(assignment.sector_id) : undefined;

    let sectorBlock: Record<string, unknown> | null = null;
    if (assignment != null) {
      sectorBlock = {
        id: assignment.sector_id,
        ja: sectorDef ? sectorDef.ja : null,
        hue: sectorDef ? sectorDef.hue : null,
        provenance: assignment.provenance,
      };
    }

    // Emit fields in EXACTLY the order Python emits them (Pydantic model_dump
    // preserves declaration order; JSON.stringify preserves object insertion
    // order). Order matters for byte-equivalence.
    const payload: Record<string, unknown> = {
      id: occId,
      schema_version: '1.2',
      title: {
        ja: occ.title_ja,
        aliases_ja: occ.aliases_ja,
      },
      classifications: {
        mhlw_main: occ.classifications.mhlw_main ?? null,
        mhlw_all: occ.classifications.mhlw_all,
        jsoc_main: occ.classifications.jsoc_main ?? null,
        jsoc_all: occ.classifications.jsoc_all,
      },
      sector: sectorBlock,
      risk_band: riskBand(score ? score.ai_risk : null),
      workforce_band: workforceBand(stats ? stats.workers ?? null : null),
      demand_band: demandBand(stats ? stats.recruit_ratio ?? null : null),
      description: {
        summary_ja: occ.description.summary_ja ?? null,
        what_it_is_ja: occ.description.what_it_is_ja ?? null,
        how_to_become_ja: occ.description.how_to_become_ja ?? null,
        working_conditions_ja: occ.description.working_conditions_ja ?? null,
      },
      ai_risk: score
        ? {
            score: score.ai_risk,
            model: score.model,
            scored_at: score.date,
            rationale_ja: score.rationale_ja,
          }
        : null,
      stats: stats
        ? {
            // Match Pydantic's exclude={"id", "schema_version"} model_dump
            source: stats.source,
            salary_man_yen: stats.salary_man_yen ?? null,
            workers: stats.workers ?? null,
            monthly_hours: stats.monthly_hours ?? null,
            average_age: stats.average_age ?? null,
            recruit_wage_man_yen: stats.recruit_wage_man_yen ?? null,
            recruit_ratio: stats.recruit_ratio ?? null,
          }
        : null,
      skills_top10: topN(occ.skills, skillsLabels, 10),
      knowledge_top5: topN(occ.knowledge, knowledgeLabels, 5),
      abilities_top5: topN(occ.abilities, abilitiesLabels, 5),
      tasks_count: occ.tasks.length,
      tasks_lead_ja: occ.tasks_lead_ja ?? null,
      related_orgs: occ.related_orgs.map((o) => ({
        name_ja: o.name_ja,
        url: o.url ?? null,
      })),
      related_certs_ja: occ.related_certs_ja,
      url: occ.url,
      data_source_versions: {
        ipd_numeric: occ.data_source_versions.ipd_numeric,
        ipd_description: occ.data_source_versions.ipd_description,
        ipd_retrieved_at: occ.data_source_versions.ipd_retrieved_at,
      },
    };

    const padded = String(occId).padStart(4, '0');
    const filePath = join(outDir, `${padded}.json`);
    await writeFile(
      filePath,
      JSON.stringify(payload) + '\n',
      'utf-8',
    );
    writtenFiles.push(filePath);
  }

  return {
    files: writtenFiles,
    dir: outDir,
    fileCount: writtenFiles.length,
  };
}
